lib/gem2rpm/rpm_dependency.rb

```diff
@@ -47,5 +47,11 @@ def to_rpm
       end
       rpm_dependencies.join("\n")
     end
+
+    # Returns string with entry suitable for RPM .spec file with RPM 4.14+.
+    def to_rich_rpm
```

Comment:
I was under the impression that `virtualize.with_requires` results in `['Requires: rubygem(mygem) >= 1.0', 'Requires: rubygem(mygem) < 2']` and if you join that, it doesn't work. I considered something like `virtualize.to_rich_rpm.with_requires.commented_out` but then `.to_rich_rpm` doesn't return a string anymore.

> Actually, second option could be to have filter, similar to how the `virtualize.with_requires.comment_out` filter chain works. Actually that would be probably the best option if it is doable

I'm not sure what you exactly mean by this. Do you mean supporting a block?

Reply:
This demonstrates how the filter chain works:

~~~
$ cat test.spec.erb 
<%
p development_dependencies
p development_dependencies.virtualize
p development_dependencies.virtualize.with_requires
p development_dependencies.virtualize.with_requires.comment_out
p development_dependencies.virtualize.with_requires.comment_out.to_rpm
-%>

$ bin/gem2rpm --template test.spec.erb test/artifacts/testing_gem/testing_gem-1.0.0.gem 
#<Gem2Rpm::RpmDependencyList:0x00007f10ee78b738 @items=[<Gem::Dependency type=:development name="test_development" requirements="~> 1.0, >= 1.0.0">]>
#<Gem2Rpm::RpmDependencyList:0x00007f10e0fa8dc0 @items=[<Gem::Dependency type=:development name="rubygem(test_development)" requirements="~> 1.0, >= 1.0.0">]>
#<Gem2Rpm::RpmDependencyList:0x00007f10e0fa88c0 @items=[<Gem::Dependency type=:development name="BuildRequires: rubygem(test_development)" requirements="~> 1.0, >= 1.0.0">]>
#<Gem2Rpm::RpmDependencyList:0x00007f10e0fa82f8 @items=[<Gem::Dependency type=:development name="# BuildRequires: rubygem(test_development)" requirements="~> 1.0, >= 1.0.0">]>
"# BuildRequires: rubygem(test_development) >= 1.0\n# BuildRequires: rubygem(test_development) < 2\n# BuildRequires: rubygem(test_development) >= 1.0.0\n"
~~~

As you can see, there still single `Gem::Dependency` object and the filters modifies the internal `name` field, keeping the `requirement` field in its original form. IOW it is still array of `Gem::Version` object.

Let me change the template a bit:

~~~
$ cat test.spec.erb 
<%
development_dependencies.virtualize.with_requires.comment_out.each {|d| p d.requirement }
-%>

$ bin/gem2rpm --template test.spec.erb test/artifacts/testing_gem/testing_gem-1.0.0.gem 
[">= 1.0", "< 2", ">= 1.0.0"]
~~~

IOW, at this stage, it is still possible to output single line or three lines, as needed. From here I continued in this direction:

~~~
$ cat test.spec.erb 
<%=
development_dependencies.virtualize.with_requires.comment_out.map do |d|
  [d.name, "(#{d.requirement.join(' with ')})"].join
end.join("\n")
-%>

$ bin/gem2rpm --template test.spec.erb test/artifacts/testing_gem/testing_gem-1.0.0.gem 
# BuildRequires: rubygem(test_development)(>= 1.0 with < 2 with >= 1.0.0)
~~~

This however unveils, that my filter idea won't work as I hoped 🤣🤦🏻‍♂️ But after all, this might be the way:

~~~
$ cat test.spec.erb 
<%=
dd = development_dependencies.virtualize.map do |d|
  Gem::Dependency.new("(#{d.requirement.map {|r| [d.name, r].join(' ')}.join(' with ')})", d.type)
end
dd = Gem2Rpm::RpmDependencyList.new dd
dd.with_requires.comment_out.to_rpm
-%>

$ bin/gem2rpm --template test.spec.erb test/artifacts/testing_gem/testing_gem-1.0.0.gem 
# BuildRequires: (rubygem(test_development) >= 1.0 with rubygem(test_development) < 2 with rubygem(test_development) >= 1.0.0)
~~~

This could be further improved if the `RpmDependencyList` implemented the `map` method or of course if there was different filter, such as `.rich_dependencies` hiding this.

Reply:
> This however unveils, that my filter idea won't work as I hoped

This is exactly what I was referring to.

I'll say that I actually like the loop in the ERB template: it makes it easier to understand what's going on. Implementing `map` (or even more parts of `Enumerable`) on `RpmDependencyList` does sound like a good idea.